### PR TITLE
enhancement: read grok aliases from file

### PR DESCRIPTION
**Pull request by itkovian**
This PR aims to address (in part) vectordotdev/vrl#89, allowing GROK aliases to be defined in one or more files. I am not sure 

- Files are assumed to contain JSON, with a single outer Object.
- Duplicate patterns are not handled
- Needs proper tests.

(recreated from https://github.com/vectordotdev/vector/pull/14914)

### CODE REVIEW

File: lib/stdlib/src/parse_groks.rs [hunk 1]
```diff
@@ -265,6 +303,31 @@ mod test {
...
+                    "%{common_prefix} %{_message}".into(),
+                    ]),
+                aliases: value!({
+                    "common_prefix": "%{_timestamp} %{_loglevel}",
+                    "_timestamp": "%{TIMESTAMP_ISO8601:timestamp}",
+                    "_loglevel": "%{LOGLEVEL:level}",
+                    "_status": "%{POSINT:status}",
+                    "_message": "%{GREEDYDATA:message}"
+                }),
+                alias_sources: Value::Array(vec![]),
```
Comment (fuchsnj) [inline lib/stdlib/src/parse_groks.rs:320]: I would like to see a test with a least 1 real alias file loaded. If it's easier, you can use the more general VRL testing framework at `lib/tests/tests/functions/...` and include a file containing an alias.

**Review by fuchsnj — COMMENTED** [review]
Added a few comments, overall it seems to be reasonable though. There was a large refactoring recently that moved most of the files from `lib/*` to `src/*`. The merge conflicts you are seeing are probably from that.

File: lib/stdlib/src/parse_groks.rs [hunk 2]
```diff
@@ -168,6 +173,39 @@ impl Function for ParseGroks {
...
+            .into_iter()
+            .map(|expr| {
+                let path = expr
+                    .resolve_constant()
+                    .ok_or(vrl::function::Error::ExpectedStaticExpression {
+                        keyword: "alias_sources",
+                        expr,
+                    })?
+                    .try_bytes_utf8_lossy()
+                    .expect("filename not bytes")
```
[the remark this answers is not in the document]
Comment (itkovian) [reply lib/stdlib/src/parse_groks.rs:188]: OK. Will change.

**Comment by itkovian** [comment]
@fuchsnj I tackled the other alias thing as well, can you check if this is what you expect to happen?

**Comment by fuchsnj** [comment]
Looks good overall. Please add a line to `CHANGELOG.md` for these changes. It also looks like the `wasm32` CI check is failing. Make sure imports or other functionality that was added that isn't supported on wasm goes in the `non_wasm` module in this file so the conditional compilation works correctly. Let me know if you need any help with this.

**Comment by pront** [comment]
Hello @itkovian, let us know if you have time to updated based on this [feedback](https://github.com/vectordotdev/vrl/pull/194#issuecomment-1609614289) or if you want to hand this off. And thanks again for submitting this!

**Comment by itkovian** [comment]
I was on holidays, will check asap.

**Comment by pront** [comment]
Hi @itkovian, there are some test failures. You can iterate on these locally by running `./scripts/checks.sh` from the repo root.

**Comment by itkovian** [comment]
Tests fail for the main branch as well?

~~~~
The following warnings were emitted during compilation:

warning: error: unable to create target: 'No available targets are compatible with triple "wasm32-unknown-unknown"'
warning: 1 error generated.
warning: error: unable to create target: 'No available targets are compatible with triple "wasm32-unknown-unknown"'
warning: 1 error generated.

error: failed to run custom build command for `zstd-sys v2.0.8+zstd.1.5.5`

Caused by:
  process didn't exit successfully: `/Users/ageorges/.cargo/build-cache/debug/build/zstd-sys-c0f0ce56419d5513/build-script-build` (exit status: 1)
  --- stdout
  cargo:rerun-if-env-changed=ZSTD_SYS_USE_PKG_CONFIG
  cargo:rustc-cfg=feature="std"
  cargo:rerun-if-changed=wasm-shim/stdlib.h
  cargo:rerun-if-changed=wasm-shim/string.h
  TARGET = Some("wasm32-unknown-unknown")
  OPT_LEVEL = Some("0")
  HOST = Some("aarch64-apple-darwin")
  cargo:rerun-if-env-changed=CC_wasm32-unknown-unknown
  CC_wasm32-unknown-unknown = None
  cargo:rerun-if-env-changed=CC_wasm32_unknown_unknown
  CC_wasm32_unknown_unknown = None
  cargo:rerun-if-env-changed=TARGET_CC
  TARGET_CC = None
  cargo:rerun-if-env-changed=CC
  CC = None
  cargo:rerun-if-env-changed=CFLAGS_wasm32-unknown-unknown
  CFLAGS_wasm32-unknown-unknown = None
  cargo:rerun-if-env-changed=CFLAGS_wasm32_unknown_unknown
  CFLAGS_wasm32_unknown_unknown = None
  cargo:rerun-if-env-changed=TARGET_CFLAGS
  TARGET_CFLAGS = None
  cargo:rerun-if-env-changed=CFLAGS
  CFLAGS = None
  cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
  CRATE_CC_NO_DEFAULTS = None
  DEBUG = Some("true")
  cargo:rerun-if-env-changed=CC_wasm32-unknown-unknown
  CC_wasm32-unknown-unknown = None
  cargo:rerun-if-env-changed=CC_wasm32_unknown_unknown
  CC_wasm32_unknown_unknown = None
  cargo:rerun-if-env-changed=TARGET_CC
  TARGET_CC = None
  cargo:rerun-if-env-changed=CC
  CC = None
  cargo:rerun-if-env-changed=CFLAGS_wasm32-unknown-unknown
  CFLAGS_wasm32-unknown-unknown = None
  cargo:rerun-if-env-changed=CFLAGS_wasm32_unknown_unknown
  CFLAGS_wasm32_unknown_unknown = None
  cargo:rerun-if-env-changed=TARGET_CFLAGS
  TARGET_CFLAGS = None
  cargo:rerun-if-env-changed=CFLAGS
  CFLAGS = None
  cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
  CRATE_CC_NO_DEFAULTS = None
  cargo:rerun-if-env-changed=CC_wasm32-unknown-unknown
  CC_wasm32-unknown-unknown = None
  cargo:rerun-if-env-changed=CC_wasm32_unknown_unknown
  CC_wasm32_unknown_unknown = None
  cargo:rerun-if-env-changed=TARGET_CC
  TARGET_CC = None
  cargo:rerun-if-env-changed=CC
  CC = None
  cargo:rerun-if-env-changed=CFLAGS_wasm32-unknown-unknown
  CFLAGS_wasm32-unknown-unknown = None
  cargo:rerun-if-env-changed=CFLAGS_wasm32_unknown_unknown
  CFLAGS_wasm32_unknown_unknown = None
  cargo:rerun-if-env-changed=TARGET_CFLAGS
  TARGET_CFLAGS = None
  cargo:rerun-if-env-changed=CFLAGS
  CFLAGS = None
  cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
  CRATE_CC_NO_DEFAULTS = None
  cargo:rerun-if-env-changed=CC_wasm32-unknown-unknown
  CC_wasm32-unknown-unknown = None
  cargo:rerun-if-env-changed=CC_wasm32_unknown_unknown
  CC_wasm32_unknown_unknown = None
  cargo:rerun-if-env-changed=TARGET_CC
  TARGET_CC = None
  cargo:rerun-if-env-changed=CC
  CC = None
  cargo:rerun-if-env-changed=CFLAGS_wasm32-unknown-unknown
  CFLAGS_wasm32-unknown-unknown = None
  cargo:rerun-if-env-changed=CFLAGS_wasm32_unknown_unknown
  CFLAGS_wasm32_unknown_unknown = None
  cargo:rerun-if-env-changed=TARGET_CFLAGS
  TARGET_CFLAGS = None
  cargo:rerun-if-env-changed=CFLAGS
  CFLAGS = None
  cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
  CRATE_CC_NO_DEFAULTS = None
  running: "clang" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "--target=wasm32-unknown-unknown" "-I" "wasm-shim/" "-I" "zstd/lib/" "-I" "zstd/lib/common" "-fvisibility=hidden" 
"-DXXH_STATIC_ASSERT=0" "-DZSTD_LIB_DEPRECATED=0" "-DXXH_PRIVATE_API=" "-DZSTDLIB_VISIBILITY=" "-DZSTDERRORLIB_VISIBILITY=" "-o" "/Users/ageorges/.cargo/build-cache/wasm32-unknown-unknown/debug/build/zstd-sys-3ee678ded
c1a7237/out/zstd/lib/common/debug.o" "-c" "zstd/lib/common/debug.c"
  cargo:warning=error: unable to create target: 'No available targets are compatible with triple "wasm32-unknown-unknown"'
  cargo:warning=1 error generated.
  exit status: 1
  running: "clang" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "--target=wasm32-unknown-unknown" "-I" "wasm-shim/" "-I" "zstd/lib/" "-I" "zstd/lib/common" "-fvisibility=hidden" 
"-DXXH_STATIC_ASSERT=0" "-DZSTD_LIB_DEPRECATED=0" "-DXXH_PRIVATE_API=" "-DZSTDLIB_VISIBILITY=" "-DZSTDERRORLIB_VISIBILITY=" "-o" "/Users/ageorges/.cargo/build-cache/wasm32-unknown-unknown/debug/build/zstd-sys-3ee678ded
c1a7237/out/zstd/lib/common/entropy_common.o" "-c" "zstd/lib/common/entropy_common.c"
  cargo:warning=error: unable to create target: 'No available targets are compatible with triple "wasm32-unknown-unknown"'
  cargo:warning=1 error generated.
  exit status: 1

  --- stderr


  error occurred: Command "clang" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "--target=wasm32-unknown-unknown" "-I" "wasm-shim/" "-I" "zstd/lib/" "-I" "zstd/lib/common" "-fvisi
bility=hidden" "-DXXH_STATIC_ASSERT=0" "-DZSTD_LIB_DEPRECATED=0" "-DXXH_PRIVATE_API=" "-DZSTDLIB_VISIBILITY=" "-DZSTDERRORLIB_VISIBILITY=" "-o" "/Users/ageorges/.cargo/build-cache/wasm32-unknown-unknown/debug/build/zst
d-sys-3ee678dedc1a7237/out/zstd/lib/common/debug.o" "-c" "zstd/lib/common/debug.c" with args "clang" did not execute successfully (status code exit status: 1).


warning: build failed, waiting for other jobs to finish...

vrl on 🌱 main [$] is 📦 v0.6.0 via 🦀 v1.72.0 took 19m55s 
✗ rustup target add wasm32-unknown-unknown
info: component 'rust-std' for target 'wasm32-unknown-unknown' is up to date
~~~~

**Comment by pront** [comment]
> Tests fail for the main branch as well?
> 
Hi @itkovian, checks are passing on the current HEAD (ad17a96f1). 

The failure you posted above seems like an environment setup issue. Two commands to try: 
```
rustup target add wasm32-unknown-unknown
cargo clean
```

If this is blocking you, you can just pick and choose what to run: 
```
./scripts/clippy.sh
./scripts/tests.sh
./scripts/vrl_tests.sh
```

**Comment by pront** [comment]
Hi @itkovian, just following up on this PR. Are you still interested in driving this to completion?

File: Cargo.toml [hunk 3]
```diff
@@ -131,7 +131,7 @@ ofb = { version = "0.6.1", optional = true }
 
 # Dependencies used for non-WASM
 [target.'cfg(not(target_arch = "wasm32"))'.dependencies]
-dns-lookup = { version = "2.0.4", optional = true }
+dns-lookup = { version = "~2.0", optional = true }
```
Comment (pront) [inline Cargo.toml:134]: What was the issue here?

Comment (itkovian) [reply Cargo.toml:134]: I'm not sure. I retried and it now builds (on my laptop :)) Will revert.

**Comment by itkovian** [comment]
@pront Yes, please :) Wasm build/tests should be fixed.

File: CHANGELOG.md [hunk 4]
```diff
@@ -47,6 +47,7 @@
 - fixed type definitions for side-effects inside of queries (https://github.com/vectordotdev/vrl/pull/258)
 - replaced `Program::final_type_state` with `Program::final_type_info` to give access to the type definitions of both the target and program result (https://github.com/vectordotdev/vrl/pull/262)
 - added `from_unix_timestamp` vrl function (https://github.com/vectordotdev/vrl/pull/277)
+- added the `alias_sources` parameter for `parse_groks` to read sources from files
```
Comment (pront) [inline CHANGELOG.md:50]: Please move this line under "unreleased".

File: src/stdlib/parse_groks.rs [hunk 5]
```diff
@@ -137,39 +144,88 @@ impl Function for ParseGroks {
...
+                        error: "alias source should be a string",
+                    })?
+                    .into_owned();
+                Ok(path)
+            })
+            .collect::<std::result::Result<Vec<String>, function::Error>>()?;
+
+        for src in alias_sources {
+            let path = Path::new(&src);
+            let file = File::open(path).map_err(|_| function::Error::InvalidArgument {
```
Comment (pront) [inline src/stdlib/parse_groks.rs:213]: I wonder if the contents of these files can be leaked. Probably not.
@fuchsnj is there any debug function that can print `src_aliases`?

File: src/stdlib/parse_groks.rs [hunk 6]
```diff
@@ -270,6 +336,31 @@ mod test {
...
+                    "%{common_prefix} %{_message}".into(),
+                    ]),
+                aliases: value!({
+                    "common_prefix": "%{_timestamp} %{_loglevel}",
+                    "_timestamp": "%{TIMESTAMP_ISO8601:timestamp}",
+                    "_loglevel": "%{LOGLEVEL:level}",
+                    "_status": "%{POSINT:status}",
+                    "_message": "%{GREEDYDATA:message}"
+                }),
+                alias_sources: Value::Array(vec![]),
```
Comment (pront) [inline src/stdlib/parse_groks.rs:353]: This test is not sufficient. See comment: https://github.com/vectordotdev/vrl/pull/194#discussion_r1207223421

**Review by pront — CHANGES_REQUESTED** [review]
Thank you @itkovian. Two things remain before we can merge this PR:

- We are missing a real test.
- I left a question about security.

Feel free to ignore the nits.

File: src/stdlib/parse_groks.rs (hunk 5 above)
Comment (fuchsnj) [reply src/stdlib/parse_groks.rs:210]: I think this is fine.

File: src/stdlib/parse_groks.rs (hunk 6 above)
Comment (itkovian) [reply src/stdlib/parse_groks.rs:350]: Added.

File: lib/tests/tests/functions/test.json [hunk 7]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (pront) [inline lib/tests/tests/functions/test.json:1]: Can you rename this to something more descriptive? E.g. `parse_groks_alias_source.json`

**Review by pront — APPROVED** [review]
Thank you for this contribution @itkovian! 

Left one comment and also some CI checks are failing (running `cargo fmt` should fix them).

**Comment by jszwedko** [comment]
Just realizing we never documented this addition in the Vector documentation. If you feel inspired, would you mind doing that @itkovian ? See https://github.com/vectordotdev/vector/pull/19081 for an example.

**Comment by itkovian** [comment]
@jszwedko Sure.

**Comment by woodsb02** [comment]
Thanks for adding this! Gentle reminder that this feature still isn't documented - glad I stumbled upon this with a google search!

**Pull request merged**